Adds support for custom BatchSampler

otaj opened the pull request:
## What does this PR do?

Adds support for custom `BatchSampler`

Fixes #13007 and #11807

### Does your PR introduce any breaking changes? If yes, please list them.

<!-- FILL IN or None -->

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

<!-- In the CHANGELOG, separate each item in the unreleased section by a blank line to reduce collisions -->

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

carmocca commented on src/pytorch_lightning/utilities/data.py:
```diff
@@ -322,12 +322,44 @@ def _dataloader_init_kwargs_resolve_sampler(
-            batch_size=batch_sampler.batch_size,
-            drop_last=(False if is_predicting else batch_sampler.drop_last),
-        )
+    batch_cls = type(batch_sampler)
```
```suggestion
    batch_sampler_cls = type(batch_sampler)
```
or
```suggestion
    sampler_cls = type(batch_sampler)
```

otaj replied:
Done

carmocca commented on src/pytorch_lightning/utilities/data.py:
```diff
@@ -322,12 +322,44 @@ def _dataloader_init_kwargs_resolve_sampler(
+            except TypeError:
+                raise MisconfigurationException(
+                    "We tried to reinstantiate your custom batch sampler and failed. "
+                    "To mitigate this, either follow API of `BatchSampler` or instantiate "
```
```suggestion
                    "To mitigate this, either follow the API of `BatchSampler` or instantiate "
```

otaj replied:
Done

carmocca commented on src/pytorch_lightning/utilities/data.py:
```diff
@@ -322,12 +322,44 @@ def _dataloader_init_kwargs_resolve_sampler(
+                    batch_size=batch_sampler.batch_size,
+                    drop_last=(False if is_predicting else batch_sampler.drop_last),
+                )
+            except TypeError:
```
This should check that the TypeError is a true positive as done in 

https://github.com/Lightning-AI/lightning/blob/daf7cec01e526c7503b99bfbf9e4c2048b8c8514/src/pytorch_lightning/utilities/data.py#L193-L202

otaj replied:
Done

carmocca commented on src/pytorch_lightning/utilities/data.py:
```diff
@@ -350,17 +382,36 @@ def _dataloader_init_kwargs_resolve_sampler(
+        kwargs[inject_name] = inject_object
+        return True, args, kwargs
+
+    return False, args, kwargs
```
What would be an example where we return False?

Shouldn't we include it as a new `kwarg` if the name isn't present already?

If the answer is no: this function should be called `_replace_value_in_saved`

otaj replied:
You're right the naming wasn't spot on, changed to your suggestion. False will be returned when we cannot find the argument in the argument list. For example, if we want to replace a sampler, but there is no sampler argument in `arg_names`, nor in `kwargs`

otaj replied:
And yes, the answer is no :smile: We should not include it in `kwargs`, because the method might not support this particular argument

carmocca commented on src/pytorch_lightning/utilities/data.py:
```diff
@@ -350,17 +382,36 @@ def _dataloader_init_kwargs_resolve_sampler(
 
 
+def _inject_arg_to_saved(
+    args: List[Any], kwargs: Dict[str, Any], arg_names: List[str], inject_name: str, inject_object: Any
```
super nit: I would put the inject argument key and value first in the signature

otaj replied:
I'm not sure I agree, but I don't have a strong preference either way, so changed according to your suggestion :smile:

carmocca commented on src/pytorch_lightning/utilities/data.py:
```diff
@@ -322,12 +322,44 @@ def _dataloader_init_kwargs_resolve_sampler(
+                if not success:
+                    rank_zero_warn(
+                        "Trying to inject `drop_last=False` into batch sampler since you are predicting, however it "
+                        f"seems the class `{batch_cls}` does not support it. Your predictions might be incomplete. "
```
The class is a class so this needs to use the qualname to avoid printing something like

`<class '__main__.Foo'>"`

```suggestion
                        f"seems the class `{batch_cls.__qualname__}` does not support it. Your predictions might be incomplete. "
```

Same comment applies to other usages of the class in a string

carmocca commented on src/pytorch_lightning/utilities/data.py:
```diff
@@ -371,18 +422,31 @@ def wrapper(obj: DataLoader, *args: Any, **kwargs: Any) -> None:
-        elif "dataset" in kwargs:
-            setattr(obj, "__dataset", kwargs["dataset"])
+        if store_explicit_args is not None:
+            for explicit_arg in store_explicit_args:
```
Is there any occurrence where we make use of the list?

otaj replied:
No. What happened was that I thought we might need some explicit args for the `BatchSampler` and had no idea how many, so changed it to list and then forgot about it to change it back :) I now made it a single string only, since there is no usecase where the list would be necessary, making it back to the list should not be a big change in the future if we ever need it

carmocca commented on src/pytorch_lightning/utilities/data.py:
```diff
@@ -350,17 +382,36 @@ def _dataloader_init_kwargs_resolve_sampler(
         dataloader.worker_init_fn = partial(pl_worker_init_function, rank=rank)
 
 
-def _wrap_dataloader_init(init: Callable) -> Callable:
```
Do you think it would be useful to keep this function name, calling _wrap_init_method where `dataset`is passed as the explicit arg?

otaj replied:
Umm... no. I don't see an added value in keeping around multiple specialized functions, when they can be made one generic function while changing one or two arguments

carmocca reviewed (commented):
This is getting really complex :joy:

Posting a partial review

otaj commented:
Thank you for the review! I addressed all your concerns :+1: 

Btw, regarding the complexity, it surprisingly wasn't too bad for me, since it's basically just logic from dataloader handling made generic

carmocca commented on src/pytorch_lightning/utilities/data.py:
```diff
@@ -371,18 +432,30 @@ def wrapper(obj: DataLoader, *args: Any, **kwargs: Any) -> None:
-            obj.__pl_dl_arg_names = param_names
+        default_kwargs = {
+            param.name: param.default
+            for param in params.values()
+            if param.name != "self"
+            and param.kind not in (param.VAR_POSITIONAL, param.VAR_KEYWORD)
```
This is very similar to the tuple created in L428. Could we create this dict first and extract the tuple from the keys of this dict?

Then you could remove the filter `(param.name not in kwargs and param.name not in param_names)`

otaj replied:
So, you're right about the fact, that it is very similar. I changed it according to your suggestion and honestly, it doesn't like like such an improvement to me. The only advantage I see is that now the intention of both of these structures is a little bit clearer.

carmocca replied:
I like this version better. Thanks!

carmocca commented on tests/tests_pytorch/utilities/test_data.py:
```diff
@@ -311,14 +313,14 @@ def test_replace_dataloader_init_method(cls, args, kwargs, arg_names, dataset, c
             assert dataloader_value is value
         else:
-            assert getattr(dataloader, key) == value
+            assert dataloader_value == value
```
why does this use `==` vs `is` based on the type?

otaj replied:
Because checking tensor for equality using == requires `all()`. And since I am keeping the reference, because the data is random, it is easier to check for object equality. Btw, this was there from previous PR. Let me know, if you think I should change it to something else.

carmocca commented on tests/tests_pytorch/utilities/test_data.py:
```diff
@@ -328,7 +330,118 @@ def test_replace_dataloader_init_method(cls, args, kwargs, arg_names, dataset, c
+            assert dataloader_value == value
+
+
+def test_replace_init_method_extra_kwargs():
```
Was this case not covered by the existing dataloader tests?

otaj replied:
Surprisingly no. And they way existing tests were set up, it couldn't even be covered, because it checked whether `__pl_saved_kwargs == kwargs`, which were the ones passed to the test. So it did not do anything with the default kwargs at all

carmocca commented on tests/tests_pytorch/utilities/test_data.py:
```diff
@@ -328,7 +330,118 @@ def test_replace_dataloader_init_method(cls, args, kwargs, arg_names, dataset, c
+            super().__init__(sampler, 10, drop_last)
+
+    with _replace_init_method(BatchSampler):
+        sampler = RandomSampler(range(10))
```
This instantiation does not need to happen inside the context manager no? Same for the DataLoader.

Same comment applies to the other tests

otaj replied:
Resolved

carmocca reviewed (approved):
Great! Minor comments

tchaton commented on src/pytorch_lightning/utilities/data.py:
```diff
@@ -322,12 +322,54 @@ def _dataloader_init_kwargs_resolve_sampler(
-            drop_last=(False if is_predicting else batch_sampler.drop_last),
-        )
+    batch_sampler_cls = type(batch_sampler)
+    if batch_sampler is not None and (batch_sampler_cls is not BatchSampler or is_predicting):
```
Hey @otaj. I am confused here. Why do we need to re0-instantiate the BatchSampler of the users ?

This was done for data loaders because they block settings arguments once instantiated. But here, we can modify it in-place.

otaj replied:
Hi @tchaton, fair question and to be honest, my best answer is "it was done like this when I came to this code". However, it might be better for custom batch samplers since they might do some setup in `__init__` based on received Sampler.

otaj commented on src/pytorch_lightning/utilities/data.py:
```diff
@@ -322,12 +322,54 @@ def _dataloader_init_kwargs_resolve_sampler(
+    batch_sampler_cls = type(batch_sampler)
+    if batch_sampler is not None and (batch_sampler_cls is not BatchSampler or is_predicting):
+        if hasattr(batch_sampler, "__pl_saved_args"):
+            args = list(batch_sampler.__pl_saved_args)
```
Because I always forget, that you can concatenate tuples, for some reason, I think of them as more primitive than they actually are. :facepalm: 

Changed to tuples

rohitgr7 reviewed (approved):
looks good! high level review.

minor comments.

carmocca commented:
@otaj There's a failing test. Also, @awaelchli would you kindly review?

awaelchli commented on src/pytorch_lightning/utilities/data.py:
```diff
@@ -322,12 +323,54 @@ def _dataloader_init_kwargs_resolve_sampler(
+            if is_predicting:
+                success, args, kwargs = _replace_value_in_saved_args("drop_last", False, args, kwargs, arg_names)
+                if not success:
+                    rank_zero_warn(
+                        f"Trying to inject `drop_last=False` into batch sampler since you are predicting, however it "
+                        f"seems the class `{batch_sampler_cls.__qualname__}` does not support it. "
+                        "Your predictions might be incomplete. To mitigate this, expose `drop_last` in the `__init__` "
+                        "method of your custom class."
+                    )
```
This is not something we do for the DataLoader I hope, right? Because there the drop_last argument is False by default. 

For the batch sampler, I don't fully understand the reasoning here. [The default value is not set](https://pytorch.org/docs/stable/data.html#torch.utils.data.BatchSampler), which means the user has to always provide the value explicitly. Why are we trying to change it then? 

So if my logic is correct here, I would strongly suggest not to interfere with the value for drop_last here, and only send out a warning.

awaelchli replied:
Edit: I just saw that we already had this logic before
```
if batch_sampler is not None and (type(batch_sampler) is not BatchSampler or is_predicting):
        batch_sampler = type(batch_sampler)(
            sampler,
            batch_size=batch_sampler.batch_size,
            drop_last=(False if is_predicting else batch_sampler.drop_last),
        )
```
I disagree with this and will open a separate issue about it. So ignore my comment for now as it does not concern this PR.

awaelchli commented on src/pytorch_lightning/utilities/data.py:
```diff
@@ -322,12 +323,54 @@ def _dataloader_init_kwargs_resolve_sampler(
+            success, args, kwargs = _replace_value_in_saved_args("sampler", sampler, args, kwargs, arg_names)
+            if not success:
+                raise MisconfigurationException(
+                    "Trying to inject modified sampler into the batch sampler; however, it seems the class "
+                    f"`{batch_sampler_cls.__qualname__}` does not support argument called sampler. To mitigate this, "
+                    "expose argument `sampler` in the `__init__` method of your custom class."
```
```suggestion
                    "Trying to inject a modified sampler into the batch sampler; however, it seems the class "
                    f"`{batch_sampler_cls.__qualname__}` does not have an argument called `sampler.` To mitigate this, "
                    "expose an argument `sampler` in the `__init__` method of your custom class."
```

otaj commented on tests/tests_pytorch/utilities/test_data.py:
```diff
@@ -328,7 +330,123 @@ def test_replace_dataloader_init_method(cls, args, kwargs, arg_names, dataset, c
+
+
+@pytest.mark.parametrize("predicting", [True, False])
+def test_custom_batch_sampler(predicting):
```
Done. Let me know what you think

otaj commented:
@awaelchli thank you very much for the review! I hope I addressed everything

otaj commented:
> @otaj There's a failing test. Also, @awaelchli would you kindly review?

@carmocca that should be taken care of now. It was a simple issue of changing the text of exception raised not matching the pytest catching

carmocca commented:
IPU failure looks flaky. I've rerun twice, going for thrice

otaj commented:
> IPU failure looks flaky. I've rerun twice, going for thrice

And it failed again. Can we maybe mark the test as flaky, or set a bit lower threshold?

awaelchli commented:
I ran it twice too. Does not look flaky to me. The environment must have changed, we may start seeing it on other PRs too. There is also the possibility that it's related to the PR. Note that in the IPU strategy we use the `_get_dataloader_init_args_and_kwargs` that was modified here. Just pointing it out as a possibility.

otaj commented:
> Note that in the IPU strategy we use the `_get_dataloader_init_args_and_kwargs` that was modified here. Just pointing it out as a possibility.

You're absolutely right regarding this, however the code should be functionally the same for normal dataloaders. But, to be fair, it doesn't seem to fail on other PRs, so it's quite likely that it is indeed a regression in this PR.

Let's try to re-run once or twice more and if it is still failing, let's investigate more

otaj commented:
There is something fishy going on. If I try to modify the `test_optimization` to force it on cpu (change accelerator, remove `RunIf` and remove `dm.trainer` checks), I pass the test every single time, and comfortably. So it really seems the issue is introduced in this PR. :facepalm:

otaj commented:
@carmocca, @awaelchli I finally managed to properly debug the cause of the IPU regression. The problem was, that I was passing all default kwargs from `torch.utils.data.DataLoader` to `poptorch.DataLoader` which didn't play nicely with it. I don't think there's a substantial change from when you reviewed it (just split `__pl_saved_kwargs` into `__pl_saved_kwargs` and `__pl_saved_default_kwargs`).

psridhar-asapp commented:
Hey @otaj I am confused on how to update my sampler to fit this API.

We are dealing with audio data and we want to batch samples by sequence length. In order to do this, we have shape files that are CSVs mapping between each audio sample and its length. Our BatchSampler reads these files, sorts utterances by length, and then constructs fixed batches of similarly long utterances such that all batches are similar in size.

I am not sure what to move to a Sampler here. We could move the loading of this shapes file to the sampler backed by a new dataset, but we do not sample from the shapes file to build batches, we consume the entire file at once.

I think part of my doubt is a lack of understanding on what exactly you do to the Sampler passed into the BatchSampler. I could move the loading of the shapes file inside a Dataset, pass in a Sampler to this dataset as a hacky way of giving the dataset to the BatchSampler, and totally ignore the Sampler in my BatchSampler logic while just reading from its member Dataset, but I imagine this will not solve the problem I'm having.

otaj commented:
Hi @psridhar-asapp. First things first, this doesn't really *add* new API. It keeps the functionality as before, just know it allows your custom `BatchSampler` to have also more arguments, that don't appear in a signature of regular `BatchSampler`.

I remember we spoke on this topic in some other issue (I think it was #11807). I don't think we have a good resolution to it yet, honestly.

We don't really do *much* to your sampler. Everything important regarding `Sampler` and `DataLoader` is happening in this method (and of course, all the other methods it calls). https://github.com/Lightning-AI/lightning/blob/caaf35689c585fba1bb33b70243e7ddd2cf5d13c/src/pytorch_lightning/trainer/connectors/data_connector.py#L238 

Currently, it sets shuffling if it was set to None, and we're training, wraps it in a distributed wrapper and does a bit more of a magic in `_update_dataloader`, which is just juggling arguments. The most important bit of call tree in `_update_dataloader` is honestly this function https://github.com/Lightning-AI/lightning/blob/caaf35689c585fba1bb33b70243e7ddd2cf5d13c/src/pytorch_lightning/utilities/data.py#L315

That is where we wrap either `BatchSampler` or `Sampler` in structures used for prediction or fault tolerance. I _think_ (and _think_ is the operative word here) it might be possible to remove the `Sampler` from being passed around and you could be happy with your `BatchSampler` only solution, but I don't really have a clear vision about that yet and it will require non-trivial engineering work.

If you don't care about either of these things, you can just expose argument `sampler` in your custom `BatchSampler`, (which needs to be a proper subclass of it and instantiated in `*_dataloader` hook) and don't use that sampler at all. I'm sorry, I don't have a better solution for you
